### ISSUE TITLE
Latest code causes sigsegv on linux...

### ISSUE THREAD

**Issue by 7oxicshadow** [reicast/reicast-emulator#678]
Just built the latest master on Linux. There are no build errors but it now crashes with the following error:

Home dir is: /home/toxicshadow/.reicast/
Linux paging: 00001000 00001000 00000FFF
evdev open: No such file or directory
MAP 00800000 w/ 25165824
MAP 20000000 w/ 25165824
MAP 04000000 w/ 16777216
MAP 06000000 w/ 16777216
MAP 0C000000 w/ 0
A8
vmem reserve: base: D4550000, aram: f4550000, vram: D8550000, ram: E0550000
Resetting mem
SIGSEGV @ F76D78A6 (fault_handler+0xEF60E576) ... F454A004 -> was not in vram
Fatal error : segfault
 in fault_handler -> ../../core/linux/common.cpp : 89 

Trace/breakpoint trap (core dumped)

**Comment by LoungeKatt** [comment]
Which game are you using and is this based on the code from the repository or does it include pull requests or any other changes not in the main branch?

**Comment by sl1pkn07** [comment]
try remove shell/linux-deps

and build with

  make \
   INCS='-I$(RZDCY_SRC_DIR) -I$(RZDCY_SRC_DIR)/deps -I$(RZDCY_SRC_DIR)/khronos' \
   LIBS='-L/usr/lib32 -lm -lrt -ldl -lGL -lpthread -lasound -lX11'

(chage /usr/lib32 for you library path. dont work with 64bits libraries )

make sre have you installed 'mesa' or your distro package with contains libgl, 'libx11' and 'alsa-libs' same like libgl)

i think the binaries included in shell/linux-deps is a quite outdated

**Comment by 7oxicshadow** [comment]
The code that I am using is from the master branch and is not modified in anyway.

The following commit breaks it:

167fba4c9fa4184b10d4445c90d212966a5a920a
"Added alsa audio backend"

The commit before that one has a header file issue and does not build but all builds leading upto that build and work as expected.

**Comment by 7oxicshadow** [comment]
The previous information I provided is invalid. I have tried building a number of revisions and have found the following:

35968c82916652944824001e24dae77835c361db
Builds and works perfectly. Full audio support and games boot as expected.
**"vmem reserve: base: CF600000, aram: ef600000, vram: D3600000, ram: DB600000"**

e3f1784a0f3c46a1e7597033e87eab5619250b45
Builds and works fine as long as you comment out lines 5 and 6 of audiostream.cpp. There is no sound as the commit suggests but everything is good with the following mem allocation:
**"vmem reserve: base: CF500000, aram: ef500000, vram: D3500000, ram: DB500000"**

f6d2e3fa7e7930af94fd284c30d3c4719fe249c0
Builds as long as you comment out line 5 of audiostream.cpp. Unfortunately this is where it breaks.
**"vmem reserve: base: D44A0000, aram: f44a0000, vram: D84A0000, ram: E04A0000"**
With the following error:
**"SIGSEGV @ F76278A6 (fault_handler+0xEF55F276) ... F449A004 -> was not in vram"**

The problem almost looks like either vram.Zero() or mem_b.Zero() fails to complete (_vmem.cpp) but I suspect that maybe the audiothread is trying to access memory that it shouldnt have access to? 

A point of interest is the vmem memory addresses. Between the one that works perfectly and the one without functioning sound there is only a 100000h offset which I assume is the original audio buffer where as the offset with the new audio backend is 4EA0000h.

Could it be that the allocated audio buffer in the new back end is too large for a 32bit system?

**Comment by 7oxicshadow** [comment]
Note: My tests were on a clean build with no modifications or branches. The emulator fails before it even attempts to load a game.

**Comment by LoungeKatt** [comment]
The only changes that weren't exclusive to android were to the audio, so the comment above sounds like the culprit. The vmem is usually a breakpoint regardless of the error.

**Comment by skmp** [comment]
I compiled and run master on fedora and rapi2 yesterday, worked alright. Can you try a clean build? What  distro are you using?

**Comment by 7oxicshadow** [comment]
@skmp I am using fedora 22 (64 bit).

I have have tried building natively and building using a 32 bit virtual machine (both fedora 22) and both result in the same error if you try to run it on the 64 bit environment.

Both builds are completely clean. I have deleted the source files and pulled them from the git master in both instances to make sure that none of the files have been altered in any way.

Other people are reporting success when building in a 64 bit environment so I can only assume that it is either a dependency (no build errors) or the program is trying to use 64 libs instead of the 32 bit ones. The only problem is there is no way of knowing where the program is falling over.

I have deleted the configs in the /home/.reicast folders and they do not get recreated when trying to run the program as it crashes before any attempts at recreating them.

If it helps I am using the NVIDIA propitiatory drivers (346.72) and have a rather old sound blaster XFI fatality sound card. I have tried launching the emulator with the export LD_LIBRARY_PATH variable set to the 32 bit lib path but that just causes it to crash when trying to access libGL.so.1 file.

Can you think of any dependency that the latest audio back end may have added?

**Comment by skmp** [comment]
I guess since now it is configurable, it depends on which audio backend you use? Which one do you use? alsa?

**Comment by skmp** [comment]
Also, can you please post the last 20-25 lines form the log? the very last line is never useful unless I have a debugger on the machine the crash happened...

**Comment by 7oxicshadow** [comment]
@skmp I cannot seem to find any log files associated with Reicast. Where would they be located?

The full output from the terminal window is:
```
[toxicshadow@UNKNOWN lin86]$ ./reicast.elf
Home dir is: /home/toxicshadow/.reicast/
Personality: 00400000
Updated personality: 00000000
Linux paging: 00001000 00001000 00000FFF
evdev open: No such file or directory
MAP 00800000 w/ 25165824
MAP 20000000 w/ 25165824
MAP 04000000 w/ 16777216
MAP 06000000 w/ 16777216
MAP 0C000000 w/ 0
A8
vmem reserve: base: D4580000, aram: f4580000, vram: D8580000, ram: E0580000
Resetting mem
SIGSEGV @ F77078A6 (fault_handler+0xEF63E556) ... F456A404 -> was not in vram
Fatal error : segfault
 in fault_handler -> ../../core/linux/common.cpp : 93 

Trace/breakpoint trap (core dumped)
[toxicshadow@UNKNOWN lin86]$ 
```

**Comment by 7oxicshadow** [comment]
Ok so I have fixed the issue but as of yet I do not understand why it makes any difference.

If I move the init_audio call before the if (!_vmem_reserve()) call then everything works fine and the emulator loads fine with sound.

If I move the the init_audio call directly after the if (!_vmem_reserve()) call it will crash in the way I mentioned in the post above.

If you ever get chance to look into why this makes a difference could you please give a brief description of why it works? I am curious as to if it is a race condition or something to do with memory initialisation....

Thanks.

**Comment by sl1pkn07** [comment]
build and run ok for me (115ee47efc71fecbbed0ac6cc5c17aac09169e22). but i need remove/unset bundled linux libs and headers (https://github.com/reicast/reicast-emulator/tree/master/shell/linux-deps) and use a system/distro ones

      sed -e 's|-I../linux-deps/include||g' \
          -e "s|../linux-deps/lib/x86|${_libs}|g" \
          -i reicast-emulator/shell/lin86/Makefile

*${libs} = 32 libs on your distro. in arch is /usr/lib32

if not make this changes, for me gets segfault

**Comment by skmp** [comment]
The linux makefile shouldn't be using the bundled stuff anyway -- it's just there for cross compiling to pandora. The crash is quite interesting, though, looks like some real bug on my code. I'll take a look

**Comment by 7oxicshadow** [comment]
Ok in order to try and assist with this bug I have done some more digging....

It turns out that if I move **any** function call (within the dc_init function) before if (!_vmem_reserve()) the emulator will work fine so it is **not** specific to audio. I tried the following:

InitAudio(); -----> emulator works as it should
webui_thd.Start(); -----> emulator works as it should
if(ParseCommandLine(argc,argv)) -----> emulator works as it should
printf("test"); -----> emulator FAILS!

The interesting thing is the memory locations that the emulator tries to use when it works against when it doesnt:

Function call infront of if (!_vmem_reserve()):

    vmem reserve: base: D3B00000, aram: f3b00000, vram: D7B00000, ram: DFB00000
    Resetting mem
    Mem alloc successful!

Unmodified Master:

    vmem reserve: base: D4590000, aram: f4590000, vram: D8590000, ram: E0590000
    Resetting mem
    SIGSEGV @ F770F8A6 (fault_handler+0xEF645F96) ... F458A004 -> was not in vram

I am starting to wonder if this is the result of compiler optimisation. I am using:

    gcc version 5.1.1 20150618 (Red Hat 5.1.1-4) (GCC)

**Comment by skmp** [comment]
Is this fc22 ?

**Comment by 7oxicshadow** [comment]
Yes all tests have been done on Fedora 22 (64 bit) with the latest updates applied

**Comment by randomstuff** [comment]
If I disable the optimizations `-O3 -frename-registers -fno-strict-aliasing -fsingle-precision-constant -ffast-math -ftree-vectorize`, I still have the same segfault (using Debian testing 64 bit compiles in 32 bit executable).

**Comment by skmp** [comment]
I just upgraded to fc22 to test, and building master without any patches, works perfectly. 

gcc (GCC) 5.1.1 20150618 (Red Hat 5.1.1-4)

**Comment by 7oxicshadow** [comment]
@skmp. Thanks for checking. It looks like it is isolated to my machine then.

I have tried clean builds in both 64 bit native and 32 bit virtual machine and neither work for me without moving the if (!_vmem_reserve()) call.

I will keep messing with it to see if I can see anything obvious from my end.

@randomstuff Have you tried moving the InitAudio() call infront of the if (!_vmem_reserve()) call? It is located in /core/nullDC.cpp as part of the dc_init() function. Would be interesting to see if you issue is the same as mine.

**Comment by randomstuff** [comment]
I tried with different versions of GCC (4.6.4-7, 4.8.4-4, 4.9.3-1, 5.1.1-14 shipped in Debian testing) but it still segfaults. Moving `InitAudio()` does not fix the issue.

**Comment by 7oxicshadow** [comment]
@randomstuff Could you post the output of the terminal window (From the point that you launch it). The  time at which it fails might be different and might help to fix your issue.

Also out of interest. Do you overclock your CPU? I realise that it should have no bearing on the issue but at the moment I am trying to find any non standard things that may have an effect. I have 2.4GHz processor overclocked to 3.4GHz.

**Comment by randomstuff** [comment]
No, I do not overclock.

<pre>Home dir is: /home/test/.reicast/
Personality: 00400000
Updated personality: 00000000
Linux paging: 00001000 00001000 00000FFF
evdev open: No such file or directory
SDK: Found 'Thrustmaster Dual Trigger 3-in-1' joystick with 8 axis and 12 buttons
Error : Unable to open file for saving 
Auto-selected audio backend "alsa" (Advanced Linux Sound Architecture).
Initializing audio backend "alsa" (Advanced Linux Sound Architecture)...
MAP 00800000 w/ 25165824
MAP 20000000 w/ 25165824
MAP 04000000 w/ 16777216
MAP 06000000 w/ 16777216
MAP 0C000000 w/ 0
A8
vmem reserve: base: D4620000, aram: f4620000, vram: D8620000, ram: E0620000
Resetting mem
SIGSEGV @ F76903E6 (fault_handler+0xEF59A79B) ... F461D304 -> was not in vram
Fatal error : segfault
 in fault_handler -> ../../core/linux/common.cpp : 93 

Trace/breakpoint trap
</pre>

**Comment by 7oxicshadow** [comment]
Well, that shoots my original theory out of the window. I am still convinced that it is something to do with timing. I have tried adding delays but that does not make any difference. Only function calls seem to have an effect.

If you are willing, you could try and experiment of moving the if (!_vmem_reserve())  statement (brakets and all) to before loadsettings();

That adds a bit more of a delay to see if it makes a difference

**Comment by randomstuff** [comment]
Indeed moving the `if(!_vmem_reserve() {…}` before `loadsettings();` fixes the issue. It now boots sucessfully.

**Comment by 7oxicshadow** [comment]
Result! At least I am not alone in this issue and that helps to prove my timing theory but for the life of me I have no idea why...... :(

Unfortunately, whilst we both have a working build at the moment. There is every chance that it will break again in future revisions. Still, make the most of it for the time being :)

**Comment by randomstuff** [comment]
Yes, it's obviously a timing issue. I didn't look at the code but this kind of fix fixes the segfault as well:

~~~cpp
int dc_init(int argc,wchar* argv[])
{
	setbuf(stdin,0);
	setbuf(stdout,0);
	setbuf(stderr,0);

        // Moved the thread start here:
	webui_thd.Start();

        // Addes this to happily wait for the thread to finish its job?
	sleep(1);

	if (!_vmem_reserve())
	{
		printf("Failed to alloc mem\n");
		return -1;
	}
~~~

**Comment by Yrij-Zhavoronkov** [comment]
../../core/./nullDC.cpp:150:9: error: ‘sleep’ was not declared in this scope
sleep(1);

without sleep() make is ok, but not effect. {(MAP1 failed 13 Failed to alloc mem)}

**Comment by 7oxicshadow** [comment]
If you want to use sleep(1); try adding the following include at the top of the file:

#include <unistd.h>

That should allow it to build correctly.

**Comment by Yrij-Zhavoronkov** [comment]
Oh.. i'm  understood that's C...

Compile.... but don't effect.

**Comment by Yrij-Zhavoronkov** [comment]
<h3>phah... it's systemd !!)) </h3>

**Comment by skmp** [comment]
I still don't quite get how this issue is triggered. You might want to try the x64 support on master (e90b17c99a046190061b3f78374b6a7d90070750 and later, lin64 folder for make). It is still far slower, but shouldn't have this kind of memory issues.

**Comment by randomstuff** [comment]
Well, on my computer it works in around 50% of runs (otherwise it segfaults).

**Comment by skmp** [comment]
Is this building for x64? or for x86?

**Comment by randomstuff** [comment]
@skmp: x86_64 does not seem to be affected.

**Comment by s1nka** [comment]
Debian Stretch x86. On last reicast git-revision, build with "cmake && make":
```
$ ./reicast 
Config dir is: /home/sinka/.reicast/
Data dir is:   /home/sinka/.reicast/
Personality: 00400000
Updated personality: 00000000
Linux paging: 00001000 00001000 00000FFF
MAP 00800000 w/ 25165824
MAP 20000000 w/ 25165824
MAP 04000000 w/ 16777216
MAP 06000000 w/ 16777216
MAP 0C000000 w/ 0
A8
vmem reserve: base: 94590000, aram: b4590000, vram: 98590000, ram: A0590000
Resetting mem
SIGSEGV @ 0xb7670516 (fault_handler+0x0xaee987d6) ... 0xb458c584 -> was not in vram
Fatal error : segfault
 in fault_handler -> core/linux/common.cpp : 99
```
